singer_sdk/_singerlib/messages.py

```diff
@@ -200,25 +198,3 @@ class ActivateVersionMessage(Message):
     def __post_init__(self) -> None:
         """Post-init processing."""
         self.type = SingerMessageType.ACTIVATE_VERSION
-
-
-def format_message(message: Message) -> str:
-    """Format a message as a JSON string.
-
-    Args:
-        message: The message to format.
-
-    Returns:
-        The formatted message.
-    """
-    return json.dumps(message.to_dict(), use_decimal=True, default=str)
-
-
-def write_message(message: Message) -> None:
-    """Write a message to stdout.
-
-    Args:
-        message: The message to write.
-    """
-    sys.stdout.write(format_message(message) + "\n")
-    sys.stdout.flush()
```

Comment:
These are arguably public API, so I don't want to remove them altogether

What do you think of inverting things here: have `SingerWriter` use these public functions.

I think we can have an intermediate generic "encoder" interface, with a default JSON implementation class and instance that is used by `format_message`:

```python
class BaseSingerMessageEncoder(ABC):
  def encode(self, message):
    ...

class JSONEncoder(BaseSingerMessageEncoder):
  ...

DEFAULT_ENCODER = JSONEncoder(...)

def format_message(message):
  return DEFAULT_ENCODER.encode(message)
```